### ISSUE TITLE
Support multiple workers per VM

### ISSUE THREAD

**Issue by eric-czech** [dask/dask-cloudprovider#173]
It may be more efficient to run many workers on larger VMs as opposed to larger numbers of worker processes on small VMs, presumably to avoid inter-VM communication.  This was a suggestion that resulted from this thread on optimizing some dask array workflows: https://github.com/pystatgen/sgkit/issues/390.

@quasiben mentioned that threads per worker can be controlled with `worker_options={"nthreads": 2 }`, but there appears to be no way to run more than one worker on a single VM.

**Comment by quasiben** [comment]
I think for this to work we need to change:

https://github.com/dask/dask-cloudprovider/blob/d5dfd99e0d29b1c6b03cb291633484fdd367ed4d/dask_cloudprovider/gcp/instances.py#L318-L330

To handle a spec which looks like the following:


```python
workers_options = {
    'worker-1': {"cls": "dask.distributed.Nanny", "opts": {"nthreads": 1}},
    'worker-2': {"cls": "dask.distributed.Nanny", "opts": {"nthreads": 2}},
}
```

**Comment by jacobtomlinson** [comment]
When running on a local machine `LocalCluster` optimises this for you right? On my laptop with 12 cores it runs three workers with four threads each.

Perhaps we can reuse this logic?

**Comment by quasiben** [comment]
You mean, have dask-cloudprovider inspect the system and assign worker/threads accordingly ?

**Comment by jacobtomlinson** [comment]
Yeah. We do this for `LocalCluster` [here](https://github.com/dask/distributed/blob/eda9bcca8baa502481bf929f39340857da1ebf09/distributed/deploy/local.py#L246-L265).

**Comment by jacobtomlinson** [comment]
dask/distributed#4377 allows us to detect and create worker/threads automatically.

```
dask-worker --nprocs=auto
```

Once a release is out we can update things here to use this new option.

**Comment by eric-valente** [comment]
any progress here?

**Comment by jacobtomlinson** [comment]
@eric-valente with the latest release of `dask` and `distributed` you should be able to set `worker_options={"nprocs": "auto"}`. Although I haven't managed to come back here and test this yet.

If you could give it a go and report back it would be much appreciated!

**Comment by eric-valente** [comment]
Hi @jacobtomlinson  Tried it with packages and no luck.   The workers will turn on in EC2 and won't connect to the scheduler and then proceed to die.  When I remove the nprocs it works fine but again only has 1 worker per VM with threads = # of cores 

worker_options={'nprocs':'auto'}

dask                      2021.4.1           pyhd8ed1ab_0    conda-forge
dask-cloudprovider        2021.3.1           pyhd8ed1ab_0    conda-forge
dask-core                 2021.4.1           pyhd8ed1ab_0    conda-forge
distributed               2021.4.1         py38h578d9bd_0    conda-forge

**Comment by jacobtomlinson** [comment]
Strange! Are you able to get any logs from the EC2 instances? They tend to get dumped in `var/log/cloud-init-output.log`.

**Comment by eric-valente** [comment]
@jacobtomlinson 

This error then worker dies:
![image](https://user-images.githubusercontent.com/35536427/117164431-77f31700-ad79-11eb-8257-8deb1d3bd6ec.png)

Here are my settings:
![image](https://user-images.githubusercontent.com/35536427/117164601-9eb14d80-ad79-11eb-8c90-7104c007f068.png)

**Comment by jacobtomlinson** [comment]
Thanks @eric-valente. You also need to set the `worker_class` kwarg to `"dask.distributed.Nanny"`.

The traceback you got above looks like a bug, I'll raise that back in distributed.

@quasiben what do you think about making `dask.distributed.Nanny` the default instead of `dask.distributed.Worker`?

**Comment by eric-valente** [comment]
Thanks again for your help here @jacobtomlinson 

Tried setting the worker class to Nanny but still same issue: 

![image](https://user-images.githubusercontent.com/35536427/117344131-edceaf80-ae59-11eb-81b4-6a05435c61bd.png)

![image](https://user-images.githubusercontent.com/35536427/117344091-e3141a80-ae59-11eb-9547-ab14577fcc21.png)

**Comment by eric-valente** [comment]
https://github.com/dask/distributed/issues/4640 seems maybe related

**Comment by eric-valente** [comment]
Seems dask.distributed.Nanny does not accept nprocs either

Fails with above init error:
--spec '{"cls": "dask.distributed.Nanny", "opts": {"nprocs": 4, "name": "dask-4ffc56fd-worker-dd6509e3"}}'

Works:
--spec '{"cls": "dask.distributed.Nanny", "opts": {"ncores": 4, "name": "dask-4ffc56fd-worker-dd6509e3"}}'

**Comment by eric-valente** [comment]
Seems like the cloud-init to add a worker uses this style of starting a worker: 
python -m distributed.cli.dask_spec tcp://x.x.x.x:8786 --spec {"cls": "distributed.nanny.Nanny", "opts": {"nprocs": 4, "name": "dask-4ffc56fd-worker-dd6509e3"}}

vs. this style:
dask-worker --nprocs=auto

**Comment by quasiben** [comment]
+1 to making `dask.distributed.Nanny` the default.

**Comment by eric-valente** [comment]
@jacobtomlinson  Yeah it seems like EC2Cluster uses python -m distributed.cli.dask_spec and passing in worker_class 

![image](https://user-images.githubusercontent.com/35536427/117379261-efb16680-ae8b-11eb-90bd-967f309b7300.png)

As suggested above, I think you might need to accept multiiple workers defined in worker_options and skip nprocs? 
```
workers_options = {
    'worker-1': {"cls": "dask.distributed.Nanny", "opts": {"nthreads": 1}},
    'worker-2': {"cls": "dask.distributed.Nanny", "opts": {"nthreads": 2}},
}

**Comment by jacobtomlinson** [comment]
> I think I could use worker_module but it is not a valid parameter for EC2Cluster

It should be, have you tried it?

**Comment by jacobtomlinson** [comment]
Right I'm with you sorry. It seems that `dask_spec` doesn't support that today. Implementing that would have to be done upstream in `distributed`.

**Comment by ansar-sa** [comment]
Could we please request / create issue in distributed upstream to integrate this work.

For my use case it makes more sense to have a few very large machines than lots of small VMs.

**Comment by jacobtomlinson** [comment]
The spec in `distributes` now accepts a list of specs. This does assume that nodes are regular as we would need to specify number of workers and cores/memory for each worker. But we should be able to implement this here now without needing any upstream changes.

**Comment by emkademy** [comment]
Are there any news on this issue?

**Comment by kumarprabhu1988** [comment]
@jacobtomlinson I tried this on `EC2Cluster` like this: 
`ec2_cluster = EC2Cluster(
    env_vars=env_vars,
    extra_bootstrap=EC2_BOOTSTRAP_COMMANDS,
    filesystem_size=cluster_config["volume_size"],
    instance_tags=cluster_config["ec2_instance_tags"],
    n_workers=cluster_config["n_workers"],
    #worker_class="distributed.nanny.Nanny", 
    worker_options={i: {"cls": Nanny, "options": {"nthreads": 2}} for i in range(4)},
    scheduler_instance_type=cluster_config["scheduler_instance_type"],
    auto_shutdown=False,
    shutdown_on_close=False,
    security=False,  # https://github.com/dask/dask-cloudprovider/issues/249,
    volume_tags=cluster_config["ec2_instance_tags"],
    worker_instance_type=cluster_config["worker_instance_type"],
)`, but this errors out with a json encode error in `VMCluster`

`~/miniconda3/envs/pricing/lib/python3.7/site-packages/dask_cloudprovider/generic/vmcluster.py in __init__(self, scheduler, worker_module, worker_class, worker_options, *args, **kwargs)
    138                             "opts": {
    139                                 **worker_options,
--> 140                                 "name": self.name,
    141                             },
    142                         }

~/miniconda3/envs/pricing/lib/python3.7/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    229         cls is None and indent is None and separators is None and
    230         default is None and not sort_keys and not kw):
--> 231         return _default_encoder.encode(obj)
    232     if cls is None:
    233         cls = JSONEncoder`

Any ideas? Will this require some changes in cloudprovider?

**Comment by jacobtomlinson** [comment]
@kumarprabhu1988 I think the `cls` need to be a string, and it probably needs to be `"distributed.nanny.Nanny"`.

**Comment by kumarprabhu1988** [comment]
@jacobtomlinson So I tried this, and no ec2 worker machine creation fails and there are no workers created.

**Comment by jacobtomlinson** [comment]
Could you set `debug=True` and grab the `/var/log/cloud-init-output.log` logs from the instances so we can see what went wrong?

**Comment by kumarprabhu1988** [comment]
Sure, I can do that. In the meanwhile I looked at the code and tried a to pass in worker configuration with some code changes and it worked. [Here's](https://github.com/dask/dask-cloudprovider/pull/389) the change I made. Let me know what you think.
